File: app/lib/analyzer/analyzers/python.ts
import type { DiscoveredFile, PartialAnalysis } from '../types';

const HTTP_METHODS = new Set(['get', 'post', 'put', 'patch', 'delete', 'options', 'head', 'websocket']);
const EXTERNALS: Record<string, string> = {
  stripe: 'Stripe', redis: 'Redis', boto3: 'AWS', openai: 'OpenAI',
  anthropic: 'Anthropic', supabase: 'Supabase', firebase_admin: 'Firebase',
  kafka: 'Kafka', aiokafka: 'Kafka', celery: 'Celery', pika: 'RabbitMQ',
  sendgrid: 'SendGrid', resend: 'Resend', twilio: 'Twilio', pymongo: 'MongoDB',
  motor: 'MongoDB', httpx: 'HTTPX', aiohttp: 'aiohttp',
};

export function analyzePython(file: DiscoveredFile): PartialAnalysis {
  const source = file.content;
  const lines = source.split(/\r?\n/);
  const lineCount = lines.length;

  const result: PartialAnalysis = {
    file: {
      id: `file:${file.path}`,
      path: file.path,
      language: 'Python',
      lines: lineCount,
      bytes: file.size,
      hash: file.hash,
      role: 'source',
      parsed: true,
      error: null,
    },
    symbols: [
      {
        id: file.path,
        type: 'module',
        name: file.path.split('/').pop()!,
        file: file.path,
        line: 1,
        end_line: lineCount,
        parent: null,
        exported: false,
        evidence: [],
      },
    ],
    imports: [],
    calls: [],
    routes: [],
    frameworks: new Set<string>(),
    databases: new Set<string>(),
    externals: new Set<string>(),
    entrypointEvidence: [],
  };

  // 1. Router variable prefixes e.g. router = APIRouter(prefix="/api/v1") or Blueprint("name", __name__, url_prefix="/api/v1")
  const routerPrefixes: Record<string, string> = {};
  const routerAssignRe = /\b([a-zA-Z_$][\w$]*)\s*=\s*(?:APIRouter|Blueprint)\s*\(([\s\S]*?)\)/g;
  let match: RegExpExecArray | null;
  while ((match = routerAssignRe.exec(source)) !== null) {
    const varName = match[1];
    const args = match[2];
    const prefixMatch = args.match(/(?:prefix|url_prefix)\s*=\s*["']([^"']+)["']/);
    if (prefixMatch) {
      routerPrefixes[varName] = prefixMatch[1];
    }
  }

  // 2. Imports: "import x, y as z" and "from .module import a, b"
  lines.forEach((lineText, idx) => {
    const lineNum = idx + 1;
    const trimmed = lineText.trim();
    if (trimmed.startsWith('#')) return;

    // from module import names
    const fromMatch = trimmed.match(/^from\s+([a-zA-Z0-9_.]+)\s+import\s+([\s\S]+)/);
    if (fromMatch) {
      const moduleName = fromMatch[1];
      const names = fromMatch[2].split(',').map((n) => n.trim().split(/\s+as\s+/)[0]).filter(Boolean);
      result.imports.push({
        id: `${file.path}:import:${lineNum}:${moduleName}`,
        source: file.path,
        module: moduleName,
        names,
        line: lineNum,
        target: null,
        external: false,
      });
      recordExternal(moduleName, result);
      return;
    }

    // import module as alias
    const importMatch = trimmed.match(/^import\s+([a-zA-Z0-9_., ]+)/);
    if (importMatch) {
      const parts = importMatch[1].split(',');
      for (const part of parts) {
        const mod = part.trim().split(/\s+as\s+/)[0];
        if (!mod) continue;
        result.imports.push({
          id: `${file.path}:import:${lineNum}:${mod}`,
          source: file.path,
          module: mod,
          names: [mod],
          line: lineNum,
          target: null,
          external: false,
        });
        recordExternal(mod, result);
      }
    }
  });

  // 3. Classes & Functions with Decorators
  let currentClass: string | null = null;
  const pendingDecorators: { name: string; args: string; line: number }[] = [];

  for (let idx = 0; idx < lines.length; idx++) {
    const lineNum = idx + 1;
    const rawLine = lines[idx];
    const trimmed = rawLine.trim();

    if (!trimmed || trimmed.startsWith('#')) continue;

    // Check indentation to know if inside class
    const indent = rawLine.search(/\S/);
    if (indent === 0 && !trimmed.startsWith('@')) {
      currentClass = null;
    }

    // Decorators: @router.get("/users") or @app.route("/items", methods=["POST"])
    if (trimmed.startsWith('@')) {
      const decMatch = trimmed.match(/^@([a-zA-Z0-9_.]+)(?:\((.*)\))?/);
      if (decMatch) {
        pendingDecorators.push({
          name: decMatch[1],
          args: decMatch[2] ?? '',
          line: lineNum,
        });
      }
      continue;
    }

    // Class definition
    const classMatch = trimmed.match(/^class\s+([a-zA-Z0-9_]+)(?:\s*\(([\s\S]*?)\))?\s*:/);
    if (classMatch) {
      const className = classMatch[1];
      const bases = (classMatch[2] ?? '').split(',').map((b) => b.trim());
      currentClass = className;

      let type = 'class';
      const evidence: string[] = [];

      const isModel = bases.some((b) =>
        b.endsWith('Base') ||
        b.endsWith('DeclarativeBase') ||
        b.includes('models.Model') ||
        b === 'Model' ||
        b.endsWith('Model') ||
        b === 'SQLModel' ||
        b.endsWith('Document')
      );

      if (isModel) {
        type = 'database_model';
        if (bases.some((b) => b.includes('models.Model'))) {
          evidence.push('inherits from Django Model');
          result.frameworks.add('Django');
          result.databases.add('Django ORM');
        } else if (bases.some((b) => b === 'SQLModel')) {
          evidence.push('inherits from SQLModel');
          result.frameworks.add('SQLModel');
          result.databases.add('SQL database');
        } else if (bases.some((b) => b.endsWith('Document'))) {
          evidence.push('inherits from Beanie/Mongo Document');
          result.databases.add('MongoDB');
        } else {
          evidence.push('inherits from an ORM model base');
          result.frameworks.add('SQLAlchemy');
          result.databases.add('SQL database');
        }
      }

      result.symbols.push({
        id: `${file.path}::${className}`,
        type,
        name: className,
        file: file.path,
        line: lineNum,
        end_line: null,
        parent: null,
        exported: true,
        evidence,
      });

      pendingDecorators.length = 0;
      continue;
    }

    // Function or method definition
    const funcMatch = trimmed.match(/^(?:async\s+)?def\s+([a-zA-Z0-9_]+)\s*\(/);
    if (funcMatch) {
      const funcName = funcMatch[1];
      const kind = currentClass ? 'method' : 'function';
      const parent = currentClass ? `${file.path}::${currentClass}` : null;
      const symbolId = currentClass ? `${file.path}::${currentClass}::${funcName}` : `${file.path}::${funcName}`;

      result.symbols.push({
        id: symbolId,
        type: kind,
        name: funcName,
        file: file.path,
        line: lineNum,
        end_line: null,
        parent,
        exported: true,
        evidence: [],
      });

      // Check pending decorators for routes
      for (const dec of pendingDecorators) {
        const decParts = dec.name.split('.');
        const methodCandidate = decParts.pop()?.toLowerCase() ?? '';
        const routerVar = decParts[0] ?? '';
        const prefix = routerPrefixes[routerVar] ?? '';

        let method = methodCandidate.toUpperCase();
        let framework = 'FastAPI';

        if (methodCandidate === 'route') {
          framework = 'Flask';
          method = 'GET';
          const methodsMatch = dec.args.match(/methods\s*=\s*\[([\s\S]*?)\]/);
          if (methodsMatch) {
            const parsedMethods = methodsMatch[1]
              .split(',')
              .map((m) => m.replace(/['"\s]/g, '').toUpperCase())
              .filter(Boolean);
            method = parsedMethods.join(',') || 'GET';
          }
        } else if (!HTTP_METHODS.has(methodCandidate)) {
          continue;
        }

        const pathMatch = dec.args.match(/["']([^"']+)["']/);
        if (!pathMatch) continue;

        const rawPath = pathMatch[1];
        let fullPath = rawPath;
        if (prefix) {
          const cleanPrefix = '/' + prefix.replace(/^\/+|\/+$/g, '');
          const cleanSub = rawPath.replace(/^\/+/, '');
          fullPath = cleanSub ? `${cleanPrefix}/${cleanSub}` : cleanPrefix;
        } else {
          fullPath = '/' + rawPath.replace(/^\/+/, '');
        }

        result.frameworks.add(framework);
        result.routes.push({
          id: `route:${file.path}:${lineNum}:${method}:${fullPath}`,
          method,
          path: fullPath,
          handler: symbolId,
          file: file.path,
          line: lineNum,
          framework,
          confidence: 0.98,
        });
      }

      pendingDecorators.length = 0;
      continue;
    }

    // Call expressions e.g. UserService.create() or save_user()
    const callMatches = trimmed.matchAll(/\b([a-zA-Z0-9_]+(?:\.[a-zA-Z0-9_]+)*)\s*\(/g);
    for (const cMatch of callMatches) {
      const callee = cMatch[1];
      if (['if', 'for', 'while', 'def', 'class', 'return', 'with', 'try', 'except'].includes(callee)) continue;
      const lastSymbol = result.symbols[result.symbols.length - 1];
      const sourceId = lastSymbol ? lastSymbol.id : file.path;

      result.calls.push({
        id: `${sourceId}:call:${lineNum}:${callee}`,
        source: sourceId,
        callee,
        file: file.path,
        line: lineNum,
        target: null,
        confidence: 0.55,
      });

      if (callee.endsWith('FastAPI')) {
        result.frameworks.add('FastAPI');
        result.entrypointEvidence.push('creates a FastAPI application');
      } else if (callee.endsWith('Flask')) {
        result.frameworks.add('Flask');
        result.entrypointEvidence.push('creates a Flask application');
      } else if (callee.endsWith('create_engine') || callee.endsWith('sessionmaker')) {
        result.frameworks.add('SQLAlchemy');
        result.databases.add('SQL database');
      }
    }

    // Execution guard
    if (trimmed.replace(/\s/g, '').includes("__name__=='__main__'") || trimmed.replace(/\s/g, '').includes('__name__=="__main__"')) {
      result.entrypointEvidence.push('contains a __main__ execution guard');
    }
  }

  const lowered = source.toLowerCase();
  if (lowered.includes('postgresql://') || lowered.includes('postgres://')) {
    result.databases.add('PostgreSQL');
  }
  if (lowered.includes('mongodb://') || lowered.includes('mongodb+srv://')) {
    result.databases.add('MongoDB');
  }

  return result;
}

function recordExternal(moduleName: string, result: PartialAnalysis) {
  const root = moduleName.split('.')[0];
  if (EXTERNALS[root]) {
    result.externals.add(EXTERNALS[root]);
  }
  if (root === 'fastapi') result.frameworks.add('FastAPI');
  else if (root === 'flask') result.frameworks.add('Flask');
  else if (root === 'django') result.frameworks.add('Django');
  else if (root === 'sqlalchemy') {
    result.frameworks.add('SQLAlchemy');
    result.databases.add('SQL database');
  } else if (root === 'sqlmodel') {
    result.frameworks.add('SQLModel');
    result.databases.add('SQL database');
  } else if (root === 'tortoise') {
    result.frameworks.add('Tortoise ORM');
    result.databases.add('SQL database');
  } else if (root === 'peewee') {
    result.frameworks.add('Peewee');
    result.databases.add('SQL database');
  } else if (root === 'beanie') {
    result.frameworks.add('Beanie');
    result.databases.add('MongoDB');
  }
}

File: core/repodna/analyzers/python.py
from __future__ import annotations

import ast
from collections.abc import Iterator

from repodna.model import CallEdge, ImportEdge, PartialAnalysis, Route, Symbol

from .base import LanguageAnalyzer


HTTP_METHODS = {"get", "post", "put", "patch", "delete", "options", "head", "websocket"}
EXTERNALS = {
    "stripe": "Stripe", "redis": "Redis", "boto3": "AWS", "openai": "OpenAI",
    "anthropic": "Anthropic", "supabase": "Supabase", "firebase_admin": "Firebase",
    "kafka": "Kafka", "celery": "Celery", "sendgrid": "SendGrid",
    "resend": "Resend", "twilio": "Twilio", "pymongo": "MongoDB",
}


def dotted_name(node: ast.AST | None) -> str:
    if isinstance(node, ast.Name):
        return node.id
    if isinstance(node, ast.Attribute):
        prefix = dotted_name(node.value)
        return f"{prefix}.{node.attr}" if prefix else node.attr
    if isinstance(node, ast.Call):
        return dotted_name(node.func)
    return ""


def literal_string(node: ast.AST | None) -> str | None:
    if isinstance(node, ast.Constant) and isinstance(node.value, str):
        return node.value
    return None


def _node_span(node: ast.AST) -> tuple[int, int]:
    return getattr(node, "lineno", 1), getattr(node, "end_lineno", getattr(node, "lineno", 1))


def _symbol_id(path: str, parents: list[str], name: str) -> str:
    return "::".join([path, *parents, name])


class _PythonVisitor(ast.NodeVisitor):
    def __init__(self, result: PartialAnalysis) -> None:
        self.result = result
        self.parents: list[str] = []
        self.symbol_stack: list[str] = []

    def visit_Import(self, node: ast.Import) -> None:
        for alias in node.names:
            self.result.imports.append(ImportEdge(
                id=f"{self.result.file.path}:import:{node.lineno}:{alias.name}",
                source=self.result.file.path,
                module=alias.name,
                names=[alias.asname or alias.name],
                line=node.lineno,
            ))
            self._record_external(alias.name)

    def visit_ImportFrom(self, node: ast.ImportFrom) -> None:
        module = "." * node.level + (node.module or "")
        names = [alias.name for alias in node.names]
        self.result.imports.append(ImportEdge(
            id=f"{self.result.file.path}:import:{node.lineno}:{module}",
            source=self.result.file.path,
            module=module,
            names=names,
            line=node.lineno,
        ))
        self._record_external(node.module or "")

    def _record_external(self, module: str) -> None:
        root = module.split(".", 1)[0]
        if root in EXTERNALS:
            self.result.externals.add(EXTERNALS[root])
        if root in {"fastapi"}:
            self.result.frameworks.add("FastAPI")
        elif root in {"flask"}:
            self.result.frameworks.add("Flask")
        elif root in {"django"}:
            self.result.frameworks.add("Django")
        elif root in {"sqlalchemy"}:
            self.result.frameworks.add("SQLAlchemy")
            self.result.databases.add("SQL database")

    def visit_ClassDef(self, node: ast.ClassDef) -> None:
        bases = {dotted_name(base) for base in node.bases}
        kind = "class"
        evidence: list[str] = []
        if any(base.endswith(("Base", "DeclarativeBase", "models.Model")) for base in bases):
            kind = "database_model"
            evidence.append("inherits from an ORM model base")
            if any("models.Model" in base for base in bases):
                self.result.frameworks.add("Django")
                self.result.databases.add("Django ORM")
            else:
                self.result.frameworks.add("SQLAlchemy")
                self.result.databases.add("SQL database")
        start, end = _node_span(node)
        symbol_id = _symbol_id(self.result.file.path, self.parents, node.name)
        self.result.symbols.append(Symbol(
            id=symbol_id,
            type=kind,
            name=node.name,
            file=self.result.file.path,
            line=start,
            end_line=end,
            parent=self.symbol_stack[-1] if self.symbol_stack else None,
            evidence=evidence,
        ))
        self.parents.append(node.name)
        self.symbol_stack.append(symbol_id)
        self.generic_visit(node)
        self.symbol_stack.pop()
        self.parents.pop()

    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
        self._visit_function(node)

    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> None:
        self._visit_function(node)

    def _visit_function(self, node: ast.FunctionDef | ast.AsyncFunctionDef) -> None:
        start, end = _node_span(node)
        kind = "method" if self.parents else "function"
        symbol_id = _symbol_id(self.result.file.path, self.parents, node.name)
        self.result.symbols.append(Symbol(
            id=symbol_id,
            type=kind,
            name=node.name,
            file=self.result.file.path,
            line=start,
            end_line=end,
            parent=self.symbol_stack[-1] if self.symbol_stack else None,
        ))
        for decorator in node.decorator_list:
            self._route_from_decorator(decorator, node, symbol_id)
        self.parents.append(node.name)
        self.symbol_stack.append(symbol_id)
        self.generic_visit(node)
        self.symbol_stack.pop()
        self.parents.pop()

    def _route_from_decorator(
        self,
        decorator: ast.AST,
        function: ast.FunctionDef | ast.AsyncFunctionDef,
        handler_id: str,
    ) -> None:
        if not isinstance(decorator, ast.Call):
            return
        decorator_name = dotted_name(decorator.func)
        method = decorator_name.rsplit(".", 1)[-1].lower()
        path = literal_string(decorator.args[0]) if decorator.args else None
        framework = "FastAPI"
        if method == "route":
            framework = "Flask"
            method = "GET"
            for keyword in decorator.keywords:
                if keyword.arg == "methods" and isinstance(keyword.value, (ast.List, ast.Tuple)):
                    methods = [literal_string(item) for item in keyword.value.elts]
                    method = ",".join(item for item in methods if item) or "GET"
        elif method not in HTTP_METHODS:
            return
        if path is None:
            return
        self.result.frameworks.add(framework)
        self.result.routes.append(Route(
            id=f"route:{self.result.file.path}:{function.lineno}:{method.upper()}:{path}",
            method=method.upper(),
            path=path,
            handler=handler_id,
            file=self.result.file.path,
            line=function.lineno,
            framework=framework,
            confidence=0.98,
        ))

    def visit_Call(self, node: ast.Call) -> None:
        name = dotted_name(node.func)
        if self.symbol_stack and name:
            self.result.calls.append(CallEdge(
                id=f"{self.symbol_stack[-1]}:call:{node.lineno}:{name}",
                source=self.symbol_stack[-1],
                callee=name,
                file=self.result.file.path,
                line=node.lineno,
            ))
        if name.endswith("FastAPI"):
            self.result.frameworks.add("FastAPI")
            self.result.entrypoint_evidence.append("creates a FastAPI application")
        elif name.endswith("Flask"):
            self.result.frameworks.add("Flask")
            self.result.entrypoint_evidence.append("creates a Flask application")
        elif name.endswith(("create_engine", "sessionmaker")):
            self.result.frameworks.add("SQLAlchemy")
            self.result.databases.add("SQL database")
        elif name.rsplit(".", 1)[-1] in {"path", "re_path"} and node.args:
            route_path = literal_string(node.args[0])
            handler_name = dotted_name(node.args[1]) if len(node.args) > 1 else "unknown"
            if route_path is not None:
                normalized_path = "/" + route_path.lstrip("/")
                self.result.frameworks.add("Django")
                self.result.routes.append(Route(
                    id=f"route:{self.result.file.path}:{node.lineno}:ANY:{normalized_path}",
                    method="ANY",
                    path=normalized_path,
                    handler=f"{self.result.file.path}::{handler_name}",
                    file=self.result.file.path,
                    line=node.lineno,
                    framework="Django",
                    confidence=0.88,
                ))
        self.generic_visit(node)

    def visit_If(self, node: ast.If) -> None:
        try:
            if ast.unparse(node.test).replace(" ", "") in {"__name__=='__main__'", '__name__=="__main__"'}:
                self.result.entrypoint_evidence.append("contains a __main__ execution guard")
        except Exception:
            pass
        self.generic_visit(node)


class PythonAnalyzer(LanguageAnalyzer):
    extensions = frozenset({".py", ".pyi"})

    def analyze(self, file, source: str) -> PartialAnalysis:
        result = PartialAnalysis(file=file)
        try:
            tree = ast.parse(source, filename=file.path, type_comments=True)
        except (SyntaxError, ValueError) as exc:
            file.error = f"parse_error:{exc.__class__.__name__}:{getattr(exc, 'lineno', 0) or 0}"
            return result
        file.parsed = True
        result.symbols.append(Symbol(
            id=file.path,
            type="module",
            name=file.path.rsplit("/", 1)[-1],
            file=file.path,
            line=1,
            end_line=file.lines,
        ))
        _PythonVisitor(result).visit(tree)
        lowered = source.lower()
        if "postgresql://" in lowered or "postgres://" in lowered:
            result.databases.add("PostgreSQL")
        if "mongodb://" in lowered or "mongodb+srv://" in lowered:
            result.databases.add("MongoDB")
        return result
